### PR TITLE
Added cdn modification for ceph 5

### PR DESCRIPTION
**Pull request by sunilangadi2**
Signed-off-by: Sunil Angadi <sangadi@redhat.com>

# Description
Fixes RHCEPHQE-905
Enabled tools repo "rhceph-5-tools-for-rhel-8-x86_64-rpms" for GA versions
To deploy ceph 5 on released GA bits.
In the cephadm args by passing **custom repo** and **custom_image** as _False_ it deploys latest GAed ceph 5 version.

Test results:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-Y3LICO
Upgrade from 5.0 cdn to 5.x latest
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-KI9NBP
upgraded from GAed ceph version  from `6.2.0-117.el8cp` to `16.2.0-143.el8cp`

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs

# Checklist:

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarin Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test

### CODE REVIEW

File: ceph/ceph_admin/__init__.py [hunk 1]
```diff
@@ -124,6 +124,15 @@ def set_tool_repo(self, repo=None):
             for node in self.cluster.get_nodes():
                 node.exec_command(sudo=True, cmd=cmd)
 
+    def set_cdn_tool_repo(self):
+        """
+        Add the cdn Tools repo on nodes.
+        """
+        cdn_repo = "rhceph-5-tools-for-rhel-8-x86_64-rpms"
```
Comment (sunilkumarn417) [inline ceph/ceph_admin/__init__.py:131]: the tools repo for 5 is in the `ceph.py` python file as well, can we use it from ceph package?

File: ceph/ceph_admin/bootstrap.py [hunk 2]
```diff
@@ -182,7 +182,7 @@ def bootstrap(self: CephAdmProtocol, config: Dict):
         if custom_repo:
```
Comment (sunilkumarn417) [inline ceph/ceph_admin/bootstrap.py:182]: can we have this from test config and applying definition to figure out what has to be set in `ceph_admin.py`.
ex., 
for custom repo, `custom_repo: "http://path_to.com/tools/repo"`
for CDN ,`custom_repo: None `(in fact we could not define this variable in test config) 

we can have discussion other possibilities.

File: ceph/ceph.py [hunk 3]
```diff
@@ -1650,6 +1650,10 @@ def setup_rhceph_cdn_repos(self, build, distro_ver):
             "rhceph-4-osd-for-rhel-8-x86_64-rpms",
         ]
 
+        repos_5x_rhel8 = [
```
Comment (sunilkumarn417) [inline ceph/ceph.py:1653]: I think, enabling 5x repo is not been used.

File: ceph/ceph_admin/bootstrap.py [hunk 4]
```diff
@@ -182,7 +182,7 @@ def bootstrap(self: CephAdmProtocol, config: Dict):
...
             self.set_tool_repo(repo=custom_repo)
         else:
-            self.set_tool_repo()
+            self.set_cdn_tool_repo()
```
Comment (sunilkumarn417) [inline ceph/ceph_admin/bootstrap.py:185]: I think we are missing out enabling the development build repositories here.
```custom_repo = "ftp://source"
if custom-repo == str: ```
Enable repo using that string will not enable dev build.

Comment (sunilkumarn417) [reply ceph/ceph_admin/bootstrap.py:185]:
``` custom_repo
	
	boolean- 
		true: use latest base_url
		false: Enable cdn repo_rpms
	string-
		enable repo with provided string
 ```

Comment (sunilangadi2) [reply ceph/ceph_admin/bootstrap.py]: Thanks @sunilkumarn417 for the logic.

**Comment by sunilangadi2** [comment]
Updated PR with review comments and used `--build released` way deploying CDN bits
ClL used 
`python run.py --v2 --osp-cred osp/osp-cred-ci-2.yaml --rhbuild 5.0 --platform rhel-8 --build released --global-conf conf/pacific/cephadm/tier1_3node_cephadm_bootstrap.yaml --suite suites/pacific/cephadm/tier_0_cephadm.yaml --inventory conf/inventory/rhel-8.4-server-x86_64.yaml --log-level DEBUG`
Test results
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-82TLIL/
Please review with new commits

File: ceph/ceph_admin/bootstrap.py [hunk 5]
```diff
@@ -179,9 +179,17 @@ def bootstrap(self: CephAdmProtocol, config: Dict):
         args = config.get("args")
         custom_repo = args.pop("custom_repo", None)
         custom_image = args.pop("custom_image", True)
+        build = config.get("build")
+
+        if build == "released":
```
Comment (sunilkumarn417) [inline ceph/ceph_admin/bootstrap.py:184]: As per our discussion, I think we thought of not using custom_image parameter from test case config.
``` custom_image = False
if build == "released" or _custom_repo.lower() ==  "cdn":
	custom_repo = "cdn"
	custom_image = False
else if _custom_repo != "cdn":
	custom_repo = base_url
	custom_image = container_image 
```

**Comment by sunilangadi2** [comment]
@psathyan after adding `build` param in config in some of the existing files same variable used
hence wherever build param used it got replaced directly with `rhbuild` 
Test results:
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-UA3BZV/
Please review and let me know you want to test any more scenarios to it.

File: ceph/ceph_admin/bootstrap.py (hunk 5 above)
Comment (sunilangadi2) [reply ceph/ceph_admin/bootstrap.py]: Yes @sunilkumarn417 later i missed out some scenarios and it's updated accordingly now
Thanks Please review.

File: ceph/ceph_admin/bootstrap.py [hunk 6]
```diff
@@ -179,9 +179,22 @@ def bootstrap(self: CephAdmProtocol, config: Dict):
...
+        base_url = config.get("base_url")
+
+        if build == "released" or custom_repo.lower() == "cdn":
+            custom_repo = "cdn"
+            custom_image = False
+        elif custom_repo != "cdn":
+            custom_repo = base_url
+            custom_image = container_image
 
         if custom_repo:
```
Comment (sunilkumarn417) [inline ceph/ceph_admin/bootstrap.py:193]: I think, we need to remove this part, custom_repo and custom image value is been already set above.

**Review by sunilkumarn417 — COMMENTED** [review]
@sunilangadi2 I think we may need a discussion, to finalize things on CDN.
cc: @psathyan

File: run.py [hunk 7]
```diff
@@ -772,6 +772,7 @@ def fetch_test_details(var) -> dict:
                 if repo.startswith("http"):
                     config["add-repo"] = repo
 
+            config["build"] = build
```
[the remark this answers is not in the document]
Comment (psathyan) [reply run.py:775]: Can we use `config["build-type"] = build` instead of just `build`

[the remark this answers is not in the document]
Comment (sunilangadi2) [reply run.py:775]: @psathyan yes that makes sense also it didn't brings any conflicts
Thanks sure will try that

File: ceph/ceph_admin/bootstrap.py (hunk 6 above)
Comment (sunilangadi2) [reply ceph/ceph_admin/bootstrap.py:193]: @sunilkumarn417 I think we need this bcz base_url and container_image is assigned when it's passed from CLI
I think you are getting confused now
Sure will discuss on that.

**Comment by sunilangadi2** [comment]
Closing this PR it got messed up with so many file changes,
Raised new PR to fix the same https://github.com/red-hat-storage/cephci/pull/934

**Pull request closed without merging**